gaeusersaccountmanager.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+)
+
+const emailHeaderKey = "X-Appengine-User-Email"
+const usernameHeaderKey = "X-Appengine-User-Nickname"
+
+type GAEUsersAccountManager struct{}
+
+func NewGAEUsersAccountManager() *GAEUsersAccountManager {
+	return &GAEUsersAccountManager{}
+}
+
+func (g *GAEUsersAccountManager) Authorize(fn AuthHTTPHandler) HTTPHandler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		// These headers are guaranteed to be present and come from AppEngine.
+		email := r.Header.Get(emailHeaderKey)
+		if email == "" {
+			// Normally a redirect to the login page is returned here, but the App Engine
+			// api takes care of that so if a request gets this far without the headers
+			// it can only be due to server misconfiguration.
+			// A general error returned here will trigger a 500 response
+			return errors.New("No authentication headers present")
+		}
+		username := r.Header.Get(usernameHeaderKey)
```

Comment:
can the same person be authenticated with two different usernames? a first time passing a nickname and a second time not passing the nickname therefore using the email?

Reply:
I don't think so, since these are google accounts and those don't let you specify a nickname. But no need to risk it, I switched to always get it from the email.